src/swarm-lang/Swarm/Language/Types.hs

```diff
@@ -566,6 +571,17 @@ substTydef (TydefInfo (Forall as ty) _) tys = foldT @t substF (fromType ty)
     Just ty' -> ty'
   substF tyF = rollT tyF
 
+-- | Replace a type alias with its definition everywhere it occurs
+--   inside a type.  Typically this is done when reporting the type of
+--   a term containing a local tydef: since the tydef is local we
+--   can't use it in the reported type.
+elimTydef :: forall t. Typical t => Var -> TydefInfo -> t -> t
```

Comment:
I wanted to test this, and I got this:
* return value
  ```c
  > def a = \p. tydef L a = rec l. a * l + Unit end; return (if p {inl (undefined: L a)} {inr ()}) end
  a True // shows type forall a. Cmd (((a * (rec l. (a * 1) + Unit)) + Unit) + Unit)
  ```
* return `undefined`
  ```c
  > def a = \p. tydef L a = rec l. a * l + Unit end; return (if p {inl (inr (): L a)} {inr ()}) end
  > a True // shows type forall a. Cmd (L a + Unit)
  ```
It also strips known type definitions from the top level.

Reply:
Hmm, strange.  The first one seems correct.  But it is bizarre that using `undefined` in the second case causes the `L` to show up in the type reported at the top level. (It still did that even when I restarted and only used the one with `undefined`.)

What do you mean by "It also strips known type definitions from the top level"?  Can you give an example?

Reply:
Sure, I mean that this should show my type, but instead, it expands it:
```c
tydef L a = rec l. a * l end
undefined: L a // not expanded
(1, undefined): L Int // shows { Int * (rec l. Int * l)
```

Reply:
Ah, I see.  That actually has nothing to do with `elimTydef`, it has to expand out the type to check that `(1, undefined)` actually has that type (whereas for `undefined` it does not need to do that). But then for some reason reports the expanded version of the type instead of your original annotation.  I will see if I can improve that.

Reply:
`(1, undefined): L Int` now shows a type of `L Int`.  It was just a simple matter of using the declared type annotation rather than the inferred type.